src/hotspot/share/opto/type.hpp

```diff
@@ -570,99 +580,179 @@ class TypeInteger : public Type {
   static const TypeInteger* minus_1(BasicType type);
 };
 
-
-
-//------------------------------TypeInt----------------------------------------
-// Class of integer ranges, the set of integers between a lower bound and an
-// upper bound, inclusive.
+/**
+ * Definition:
+ *
+ * A TypeInt represents a set of jint values. An jint v is an element of a
+ * TypeInt iff:
+ *
+ * v >= _lo && v <= _hi && juint(v) >= _ulo && juint(v) <= _uhi && _bits.is_satisfied_by(v)
+ *
+ * Multiple set of parameters can represent the same set.
+ * E.g: consider 2 TypeInt t1, t2
+ *
+ * t1._lo = 2, t1._hi = 7, t1._ulo = 0, t1._uhi = 5, t1._bits._zeros = 0x0, t1._bits._ones = 0x1
+ * t2._lo = 3, t2._hi = 5, t2._ulo = 3, t2._uhi = 5, t2._bits._zeros = 0xFFFFFFF8, t2._bits._ones = 0x1
+ *
+ * Then, t1 and t2 both represent the set {3, 5}. We can also see that the
+ * constraints of t2 are optimal. I.e there exists no TypeInt t3 which also
+ * represents {3, 5} such that:
+ *
+ * t3._lo > t2._lo || t3._hi < t2._hi || t3._ulo > t2._ulo || t3._uhi < t2._uhi ||
+ *     (t3._bits._zeros & t2._bis._zeros) != t3._bits._zeros || (t3._bits._ones & t2._bits._ones) != t3._bits._ones
+ *
+ * The last 2 conditions mean that the bits in t3._bits._zeros is not a subset
+ * of those in t2._bits._zeros, the same applies to _bits._ones
+ *
+ * As a result, every TypeInt is canonicalized to its optimal form upon
+ * construction. This makes it easier to reason about them in optimizations.
+ * E.g a TypeInt t with t._lo < 0 will definitely contain negative values. It
+ * also makes it trivial to determine if a TypeInt instance is a subset of
+ * another.
+ *
+ * Properties:
+ *
+ * 1. Since every TypeInt instance is canonicalized, all the bounds must also
+ * be elements of such TypeInt. Or else, we can tighted the bounds by narrowing
+ * it by one, which contradicts the assumption of the TypeInt being canonical.
+ *
+ * 2. Either _lo == jint(_ulo) and _hi == jint(_uhi), or all elements of a
+ * TypeInt lie in the intervals [_lo, jint(_uhi)] or [jint(_ulo), _hi]
+ *
+ * Proof: For 2 jint value x, y such that they are both >= 0 or < 0. Then:
+ *
+ * x <= y iff juint(x) <= juint(y)
+ *
+ * Then, we have:
+ *
+ * For a TypeInt t, there are 3 possible cases:
+ *
+ * a. t._lo >= 0. Since 0 <= t._lo <= jint(t._ulo), we have:
+ *
+ * juint(t._lo) <= juint(jint(t._ulo)) == t._ulo <= juint(t._lo)
```

Comment:
You should say what steps you are applying here... otherwise the reader has a lot to do.
Lemma, return-cast, `t._lo <= jint(t._ulo)` (maybe its own Lemma2?)